File: frontend/src/lib/types/smart-report.ts
/**
 * Structured Smart Report types — mirror the backend Pydantic models in
 * `backend/app/schemas/smart_report.py`. The backend parses the LLM's JSON
 * payload out of the stored response string and ships it as a first-class
 * `report` object, so every field is optional: a tolerant parser means a
 * partial or slightly-off model output still renders.
 */

export interface LabParameterValue {
  date?: string | null;
  value?: string | null;
}

export type ParameterStatus = "in_range" | "out_of_range" | "borderline" | "critical" | string;
export type Trend =
  | "improved"
  | "further_decreased"
  | "stable"
  | "new_abnormal"
  | "not_available"
  | string;
export type SystemStatus = "needs_attention" | "ideal" | "no_data" | string;
export type Priority = "high" | "medium" | "low" | string;

export interface LabParameter {
  name?: string | null;
  value?: string | null;
  unit?: string | null;
  date?: string | null;
  status?: ParameterStatus | null;
  reference_range?: string | null;
  trend?: Trend | null;
  previous_values?: LabParameterValue[];
}

export interface SystemGlance {
  system?: string | null;
  status?: SystemStatus | null;
  summary?: string | null;
  parameters_total?: number;
  parameters_out_of_range?: number;
  parameters_improved?: number;
}

export interface OrganDetail {
  system?: string | null;
  parameters?: LabParameter[];
}

export interface ParameterInFocus {
  name?: string | null;
  system?: string | null;
  explanation?: string | null;
  significance?: string | null;
  trend_note?: string | null;
  recommendation?: string | null;
}

export interface SmartRecommendation {
  category?: string | null;
  priority?: Priority | null;
  action?: string | null;
  reasoning?: string | null;
}

export interface SmartReportData {
  systems_at_a_glance?: SystemGlance[];
  organ_details?: OrganDetail[];
  parameters_in_focus?: ParameterInFocus[];
  recommendations?: SmartRecommendation[];
  /** Tolerate extra LLM-emitted keys (backend uses extra="allow"). */
  [key: string]: unknown;
}

File: backend/app/schemas/smart_report.py
"""Pydantic models for the AI Smart Report structured payload.

Mirrors the JSON schema requested by ``SMART_REPORT_PROMPT``
(``app/prompts/insight_prompts.py``). The models are intentionally tolerant:
collections default to empty, nested fields are optional, enums accept any
string, and ``coerce_numbers_to_str`` lets a numeric LLM value still parse.
This keeps a slightly-off model output from invalidating the whole report —
the viewer degrades to prose only when parsing fails entirely.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


class LabParameterValue(BaseModel):
    """A single prior measurement for a lab parameter."""

    model_config = ConfigDict(extra="ignore", coerce_numbers_to_str=True)
    date: str | None = None
    value: str | None = None


class LabParameter(BaseModel):
    """One lab test result with trend history."""

    model_config = ConfigDict(extra="allow", coerce_numbers_to_str=True)
    name: str | None = None
    value: str | None = None
    unit: str | None = None
    date: str | None = None
    # in_range | out_of_range | borderline | critical
    status: str | None = None
    reference_range: str | None = None
    # improved | further_decreased | stable | new_abnormal | not_available
    trend: str | None = None
    previous_values: list[LabParameterValue] = Field(default_factory=list)


class SystemGlance(BaseModel):
    """Body-system roll-up shown in the "at a glance" grid."""

    model_config = ConfigDict(extra="allow", coerce_numbers_to_str=True)
    system: str | None = None
    # needs_attention | ideal | no_data
    status: str | None = None
    summary: str | None = None
    parameters_total: int = 0
    parameters_out_of_range: int = 0
    parameters_improved: int = 0


class OrganDetail(BaseModel):
    """Lab parameters grouped by body system."""

    model_config = ConfigDict(extra="allow")
    system: str | None = None
    parameters: list[LabParameter] = Field(default_factory=list)


class ParameterInFocus(BaseModel):
    """An abnormal / noteworthy parameter with clinical context."""

    model_config = ConfigDict(extra="allow", coerce_numbers_to_str=True)
    name: str | None = None
    system: str | None = None
    explanation: str | None = None
    significance: str | None = None
    trend_note: str | None = None
    recommendation: str | None = None


class SmartRecommendation(BaseModel):
    """A clinical recommendation with priority and rationale."""

    model_config = ConfigDict(extra="allow", coerce_numbers_to_str=True)
    category: str | None = None
    # high | medium | low
    priority: str | None = None
    action: str | None = None
    reasoning: str | None = None


class ChronicCondition(BaseModel):
    """An existing/chronic condition surfaced from the full visit history."""

    model_config = ConfigDict(extra="allow", coerce_numbers_to_str=True)
    name: str | None = None
    # active | resolved | monitoring
    status: str | None = None
    since: str | None = None
    note: str | None = None


class SmartReportData(BaseModel):
    """Top-level Smart Report payload."""

    model_config = ConfigDict(extra="allow")
    chronic_conditions: list[ChronicCondition] = Field(default_factory=list)
    systems_at_a_glance: list[SystemGlance] = Field(default_factory=list)
    organ_details: list[OrganDetail] = Field(default_factory=list)
    parameters_in_focus: list[ParameterInFocus] = Field(default_factory=list)
    recommendations: list[SmartRecommendation] = Field(default_factory=list)
